### ISSUE TITLE
Design a singularity spectrum code

### ISSUE THREAD

**Issue by LRydin** [LRydin/MFDFA#2]
Include a Singularity spectrum code to understand the strength of the multifractality of the timeseries:

![Screenshot from 2020-07-03 17-57-57](https://user-images.githubusercontent.com/47757762/86484600-dc5ad980-bd56-11ea-8606-92fcf3b92d4e.png)
 
This is found in: [Multifractal detrended fluctuation analysis of nonstationary time series](https://doi.org/10.3389/fphys.2012.00141), or for example on the Matlab code and paper: [Introduction to multifractal detrended fluctuation analysis in Matlab](https://www.sciencedirect.com/science/article/abs/pii/S0378437102013833?via%3Dihub)

**Comment by mikemelon** [comment]
I don't know if it's appropriate to ask this question here, but I want to know, the bell curve of $f(\alpha)-\alpha$ is generally opening downward according to my experience in reading related articles, but for the experimental  picture below, the opening of my picture below is upward. I noticed that the small variation of of H(q)--q , is it caused this problem or  is there something wrong with my MF-DFA calculation that prevents it from converging. Thank you for notice this question.
![My experimental result picture](https://github.com/mikemelon/LYQ-notes/blob/main/images/2022-09-22_153623.png)

**Comment by LRydin** [comment]
Hey @mikemelon that is indeed strange. You are absolutely correct in your observation. `H(q)` should be a constant or decreasing function of `q`, mathematically. But it is not so unusual that in real-world data this is not the case. Could you show me how you define `q` and how you get these plots out?

Cheers!

**Comment by mikemelon** [comment]
@LRydin Thank you. you are too quick,  I define q in np.linspace(-5,5, 11), however, the adjustion of q is not bother,  the modification of length of data limits does! 
Here is some of my code, I divided MFDFA in two steps, and plot H(q)-q  and $f(\alpha)-\alpha$ in an iteration.
btw, I have got even confused results, see picture below.
![my code](https://github.com/mikemelon/LYQ-notes/blob/main/images/2022-09-22_164009.png?raw=true)
![my pic2](https://github.com/mikemelon/LYQ-notes/blob/main/images/2022-09-22_164058.png?raw=true)

**Comment by LRydin** [comment]
Hey @mikemelon, this is indeed strange. Some of the results make sense, e.g., the blue curve that has `f(α)` below `1`. strangely this is the negative side of the `q` values, which is usually where the results are less accurate.

Could you check one thing for me? Can you run:
```
pip install --upgrade MFDFA
```
just to make sure you have the latest package?

Besides that, what could lead to these results, in my opinion, is data that is too short or the `limits` that you are using being too close to each other or too close to `0`... Can you fiddle around with that and see what comes out?

Cheers!

**Comment by mikemelon** [comment]
@LRydin  Yes I have updated to the newest version. I replaced order=1 with order=4  to 7  in calculation of Fq(s), the situation is better.  After that, some of the $f(\alpha)-\alpha$ curves are not distorted , but there are still some $f(\alpha)-\alpha$ curves that open upward. I implemented MFDFA algorithm on my own independently and it remains the same。
The length of the series does not matter, because I mannually repeat the series twice or more,  everything remains  unchanged.
 I changed the limits either it remains the same. 
I also use the guide in  {Espen A. F. Ihlen, 2012, Introduction to Multifractal Detrended Fluctuation Analysis in Matlab} using MATLAB step by step but the results are similiar... (There are two implemention of MFDFA in this article I have tried the first one)
I think the main reason perhaps lies in the slope fitted in the Fq(s) calculation, as you point out, the H(q) should decline when q increases. The strange distored curves do not follow this rule.
Can it be better to use EMD or Wavelet in MF-DFA implemention, or do you have any advice , direction, or articles, please.  Thank u in advance.
BTW, how can I find your contact info (email, etc), mine is mikemelon@sina.com 
Best wishes!

**Comment by LRydin** [comment]
Hey @mikemelon, my email is hidden in the [contributions.md](https://github.com/LRydin/MFDFA/blob/master/contributions.md), but I'm writing to you directly from my email. If you have access to MATLAB, then I definitely suggest trying out a wavelet analysis, as that is an equivalently powerful tool to analyse multifractality. If you are seeing similar results this likely points to some details in your data, e.g., low precision or being too short, but besides that, I can't really think of anything else :/.

**Comment by settwi** [comment]
Is the singularity spectrum code implemented? I would like to compare it with a WTMM analysis.
Regards
William

**Comment by LRydin** [comment]
Hey @settwi, sorry for the delay, I was absent this past week. Yes, the singularity spectrum should be working (there might be some kinks) right [here (L.32)](https://github.com/LRydin/MFDFA/blob/ef652e12174541e5b2ad96260b919b40f8b99d9e/MFDFA/singspect.py#L23)

To get it to work, run

```python
from MFDFA import MFDFA
from MFDFA.singspect import singularity_spectrum
```

then run your MFDFA analysis and use that function. There is also a plot helper (if you are using `matplotlib`) called `singularity_spectrum_plot` at [L307](https://github.com/LRydin/MFDFA/blob/ef652e12174541e5b2ad96260b919b40f8b99d9e/MFDFA/singspect.py#L307)

Check if that helps!
